### PR TITLE
Test nim inference workflow and make minor updates

### PR DESCRIPTION
**Pull request by ktangsali**
# PhysicsNeMo-CFD Pull Request

## Description

Tested the nim inference workflow and made minor edits as a part of the testing. The l2_error_vs_sdf now clip first and then compute the metrics. Also, device is handled better. 

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo-cfd/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo-cfd/blob/main/CHANGELOG.md) is up to date with these changes.

### CODE REVIEW

File: physicsnemo/cfd/postprocessing_tools/metrics/l2_errors.py [hunk 1]
```diff
@@ -14,11 +14,23 @@
...
 
+import os
+import time
+
 import numpy as np
 import pyvista as pv
 import torch
 from physicsnemo.nn.functional import signed_distance_field
 
+
```
Comment (peterdsharpe) [inline physicsnemo/cfd/postprocessing_tools/metrics/l2_errors.py:30]: It looks like this accidentally leaves in debug code?

File: physicsnemo/cfd/postprocessing_tools/metrics/l2_errors.py [hunk 2]
```diff
@@ -242,12 +254,54 @@ def compute_error_vs_sdf(
...
+            f"query_points.device={query_points.device}",
+            flush=True,
+        )
+
+    t0 = time.perf_counter()
     sdf_field, _ = signed_distance_field(
         stl_vertices, stl_indices, query_points, use_sign_winding_number=True
     )
-    sdf_field = sdf_field.numpy()
+    if _debug_sdf():
```
Comment (peterdsharpe) [inline physicsnemo/cfd/postprocessing_tools/metrics/l2_errors.py]: This debug code should be cleaned up (i.e., removed)

File: physicsnemo/cfd/postprocessing_tools/metrics/l2_errors.py [hunk 3]
```diff
@@ -242,12 +254,54 @@ def compute_error_vs_sdf(
...
+        spatial_mask = None
+        query_points_np = points_all
+
+    n_all = int(points_all.shape[0])
+    n_q = int(query_points_np.shape[0])
+    # Warp launch device follows *torch tensor device* (see FunctionSpec.warp_launch_context).
+    # Default torch.tensor(...) is CPU → SDF would run on CPU and look "stuck" for large N.
+    torch_device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+    query_points = torch.tensor(query_points_np, dtype=torch.float32, device=torch_device)
+    stl_vertices = stl_vertices.to(torch_device)
```
Comment (peterdsharpe) [inline physicsnemo/cfd/postprocessing_tools/metrics/l2_errors.py]: Why define `stl_vertices` above as a torch.tensor and then transfer it here? Would be much cleaner to just do it in one shot. Same for other variables throughout the diff in this file.

**Review by peterdsharpe — CHANGES_REQUESTED** [review]
Submitting initial comments on everything except the notebooks (will review locally).

Other than the comments here, is it possible to save the PNG files (specifically `resampled_volume_errors.png` and `variations_due_to_checkpoint.png`) in a smaller file size? Right now those are adding ~25 MB to the repo size in total. Maybe JPEG or lower-res PNG?

File: physicsnemo/cfd/postprocessing_tools/metrics/l2_errors.py (hunk 3 above)
Comment (ktangsali) [reply physicsnemo/cfd/postprocessing_tools/metrics/l2_errors.py]: Good point, done.

File: physicsnemo/cfd/postprocessing_tools/metrics/l2_errors.py (hunk 2 above)
Comment (ktangsali) [reply physicsnemo/cfd/postprocessing_tools/metrics/l2_errors.py]: Done

File: physicsnemo/cfd/postprocessing_tools/metrics/l2_errors.py (hunk 1 above)
Comment (ktangsali) [reply physicsnemo/cfd/postprocessing_tools/metrics/l2_errors.py:30]: Done

**Comment by ktangsali** [comment]
> Submitting initial comments on everything except the notebooks (will review locally).
> 
> Other than the comments here, is it possible to save the PNG files (specifically `resampled_volume_errors.png` and `variations_due_to_checkpoint.png`) in a smaller file size? Right now those are adding ~25 MB to the repo size in total. Maybe JPEG or lower-res PNG?

Thanks! Compressed the images.

File: .github/workflows/ci-precommit.yml [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+          uv venv --clear --python ${{ matrix.python-version }}
+          uv pip install --python .venv/bin/python pre-commit
+          echo "$PWD/.venv/bin" >> "$GITHUB_PATH"
+      # Mirror every check-style Makefile target so the Makefile remains the
+      # single source of truth. Skip ``install`` / ``editable-install`` /
+      # ``setup-ci`` (env bootstrap done above), ``pytest`` (own workflow in
+      # ``ci-tests.yml``)
+      - name: Run pre-commit checks via Makefile
+        run: |
+          make black
```
Comment (peterdsharpe) [inline .github/workflows/ci-precommit.yml:36]: Why does this use the Makefile rather than `uv run --no-sync pre-commit run -a`?

Also, rather than manually installing pre-commit here, we should probably be installing this as a uv dev dependency in `pyproject.toml` (similar to how we do it in the core repo)

Comment (ktangsali) [reply .github/workflows/ci-precommit.yml:36]: So the makefile is still used by Blossom to run its CI. This way we can keep both those commands in sync. Adding it as dev dependency would also mean we install pnemo-cfd for this. I feel that is not required for running these linting tests

File: physicsnemo/cfd/postprocessing_tools/metrics/l2_errors.py [hunk 5]
```diff
@@ -238,16 +239,41 @@ def compute_error_vs_sdf(
         - "bin_edges": List of SDF bin edges
         - "mean_errors": List of mean errors for each SDF bin
     """
+    points_all = np.asarray(data.points)
```
Comment (peterdsharpe) [inline physicsnemo/cfd/postprocessing_tools/metrics/l2_errors.py]: This code has a latent bug:

`spatial_mask` and `sdf_field` have length `n_points` (or its `bounds`-filtered subset). When `dtype="cell"`, `data.get_array(true, preference="cell")` returns length `n_cells`, so `true_field[spatial_mask]` will raise `IndexError` (when `n_points > n_cells`) or, worse, silently slice the wrong rows (when `n_points <= n_cells`, the mask just truncates). The notebook only exercises `dtype="point"` so this is latent, but the new code shape makes it the natural thing to break next.

The bug existed in the prior version too (it computed `points = data.points` and masked the field with that). Since the refactor rewrites this exact block, fix it now: when `dtype == "cell"`, build the mask against `data.cell_centers().points` and use cell centers as the SDF query points; when `dtype == "point"`, keep current behavior.

Better still, lift the duplicated mask/`field_type` boilerplate out of `compute_l2_errors` and `compute_error_vs_sdf` into a helper: both functions repeat the same `points/bounds → bool mask → field_type detection` pattern.

Comment (peterdsharpe) [reply physicsnemo/cfd/postprocessing_tools/metrics/l2_errors.py]: Separately, if we can type-hint this function, that would be great. Specifically, including a `Literal` type for `dtype` would be helpful, as the naming convention is counterintuitive - usually `dtype` indicates array types (e.g., fp32, fp64, int64), not semantic meaning ("point" vs. "cell"). This diff technically doesn't touch the signature, but it'd be a nice-to-have for readers.

File: physicsnemo/cfd/postprocessing_tools/metrics/l2_errors.py [hunk 6]
```diff
@@ -238,16 +239,41 @@ def compute_error_vs_sdf(
...
+            & (points_all[:, 2] <= bounds[5])
+        )
+        query_points_np = points_all[spatial_mask]
+    else:
+        spatial_mask = None
+        query_points_np = points_all
+
+    # Warp launch device follows the torch tensor device (see FunctionSpec.warp_launch_context),
+    # so build all SDF inputs directly on the target device to avoid a slow CPU fallback.
+    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```
Comment (peterdsharpe) [inline physicsnemo/cfd/postprocessing_tools/metrics/l2_errors.py]: Many other functions in this repo (e.g., [interpolate_mesh_to_pc](physicsnemo/cfd/postprocessing_tools/interpolation/interpolate_mesh_to_pc.py)) accept `device: str = "cpu"` (and map `"gpu"` → `"cuda"` for back-compat). `compute_error_vs_sdf` should match that signature. As written, on a multi-GPU host it always lands on `cuda:0` with no override. Suggested signature: `device: str = "cpu"` (defaulting to CPU is consistent with the rest of `postprocessing_tools` and keeps the function pure for users without CUDA).

File: physicsnemo/cfd/postprocessing_tools/metrics/l2_errors.py [hunk 7]
```diff
@@ -238,16 +239,41 @@ def compute_error_vs_sdf(
         - "bin_edges": List of SDF bin edges
         - "mean_errors": List of mean errors for each SDF bin
     """
+    points_all = np.asarray(data.points)
+    # Apply bounds *before* SDF: evaluating winding-number SDF on full-volume meshes
```
Comment (peterdsharpe) [inline physicsnemo/cfd/postprocessing_tools/metrics/l2_errors.py]: Related to bug in the comment immediately above this line: this function has no unit tests, which would have caught the issue above.

**Review by peterdsharpe — CHANGES_REQUESTED** [review]
Overall looks good. Leaving my notebook comments here, since GitHub won't let me add line-by-line annotations within *.ipynb's.

-----

General items:
- Please update changelog (warp version bump), Dependencies section.

-----

A blocker in our notebooks, relating to us pointing users to deprecated APIs:
- The cell that runs NIM inference uses `_create_nbrs_surface` and `_interpolate`. These are explicitly deprecated in the same file, and we direct users to use ``interpolate_mesh_to_pc`` or ``physicsnemo.nn.functional.neighbors.knn`` instead. Can we fix our tutorial code here to use non-deprecated APIs? The NIM produces a sparse point cloud `output_dict["surface_coordinates"][0]` with fields. Wrap it as `pv.PolyData(coords)` with `point_data` populated, then call [interpolate_mesh_to_pc](physicsnemo/cfd/postprocessing_tools/interpolation/interpolate_mesh_to_pc.py) with `pc=mesh_cell_centers, mesh=nim_pc, mesh_dtype="point", device="gpu"`. About 5 lines of glue.

-----

Another blocker in our notebooks, related to credentials leaking:

The cells that run `!wget "{url}"` for `drivaer_418.stl` and `drivaer_420.stl` capture the redirect chain, including:

```
https://cas-bridge.xethub.hf.co/xet-bridge-us/.../drivaer_418.stl?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=cas%2F...&X-Amz-Signature=155cd1bfd12bedc99e9a5c8eafca8602bb1edbbf108495138d9d0af5906e971d&...
```

These are pre-signed S3-style URLs. They expired ~1 hour after generation (so this PR's URLs are dead now), but committing them is bad hygiene: it normalizes embedding presigned URLs in public artifacts, and future re-runs will commit live ones. Concrete fixes, in order of preference:

1. Replace `!wget "{url}" -O "{filename}"` with `huggingface_hub.hf_hub_download` (already a transitive dep via `evaluation-hf`). It logs nothing comparable.
2. Use `requests.get(url, stream=True)` with `iter_content` and a small progress bar that doesn't echo the URL.
3. At minimum, add `--quiet` to `wget` and clear the output cells before commit.

File: .github/workflows/ci-precommit.yml (hunk 4 above)
Comment (peterdsharpe) [reply .github/workflows/ci-precommit.yml:36]: That's fair. Can we instead do both? I.e.,:

- Keep this file `ci-precommit.yml` and the Makefile as you have it, but also
- Add `pre-commit` as a dev dependency in `pyproject.toml`?

As it stands, a local developer doesn't automatically get `pre-commit`, while they should - this will improve the dev experience. (This wouldn't affect CI lint speed, since as you said, CI doesn't install the package.)

Comment (peterdsharpe) [reply .github/workflows/ci-precommit.yml:36]: Separately - currently in this file, the `pre-commit` version is unpinned, effectively making CI a moving target. Let's at least hard-pin this here so we don't get weird sudden failures in the future.

File: physicsnemo/cfd/postprocessing_tools/metrics/l2_errors.py (hunk 7 above)
Comment (ktangsali) [reply physicsnemo/cfd/postprocessing_tools/metrics/l2_errors.py]: Good call on unittests, added.

File: physicsnemo/cfd/postprocessing_tools/metrics/l2_errors.py (hunk 6 above)
Comment (ktangsali) [reply physicsnemo/cfd/postprocessing_tools/metrics/l2_errors.py]: Good call, updated.

File: physicsnemo/cfd/postprocessing_tools/metrics/l2_errors.py (hunk 5 above)
Comment (ktangsali) [reply physicsnemo/cfd/postprocessing_tools/metrics/l2_errors.py]: Done!

File: .github/workflows/ci-precommit.yml (hunk 4 above)
Comment (ktangsali) [reply .github/workflows/ci-precommit.yml:36]: I am assuming you are ok with this, resolving.

**Comment by ktangsali** [comment]
> Overall looks good. Leaving my notebook comments here, since GitHub won't let me add line-by-line annotations within *.ipynb's.
> 
> General items:
> 
> * Please update changelog (warp version bump), Dependencies section.
> 
> A blocker in our notebooks, relating to us pointing users to deprecated APIs:
> 
> * The cell that runs NIM inference uses `_create_nbrs_surface` and `_interpolate`. These are explicitly deprecated in the same file, and we direct users to use `interpolate_mesh_to_pc` or `physicsnemo.nn.functional.neighbors.knn` instead. Can we fix our tutorial code here to use non-deprecated APIs? The NIM produces a sparse point cloud `output_dict["surface_coordinates"][0]` with fields. Wrap it as `pv.PolyData(coords)` with `point_data` populated, then call [interpolate_mesh_to_pc](physicsnemo/cfd/postprocessing_tools/interpolation/interpolate_mesh_to_pc.py) with `pc=mesh_cell_centers, mesh=nim_pc, mesh_dtype="point", device="gpu"`. About 5 lines of glue.
> 
> Another blocker in our notebooks, related to credentials leaking:
> 
> The cells that run `!wget "{url}"` for `drivaer_418.stl` and `drivaer_420.stl` capture the redirect chain, including:
> 
> ```
> https://cas-bridge.xethub.hf.co/xet-bridge-us/.../drivaer_418.stl?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=cas%2F...&X-Amz-Signature=155cd1bfd12bedc99e9a5c8eafca8602bb1edbbf108495138d9d0af5906e971d&...
> ```
> 
> These are pre-signed S3-style URLs. They expired ~1 hour after generation (so this PR's URLs are dead now), but committing them is bad hygiene: it normalizes embedding presigned URLs in public artifacts, and future re-runs will commit live ones. Concrete fixes, in order of preference:
> 
> 1. Replace `!wget "{url}" -O "{filename}"` with `huggingface_hub.hf_hub_download` (already a transitive dep via `evaluation-hf`). It logs nothing comparable.
> 2. Use `requests.get(url, stream=True)` with `iter_content` and a small progress bar that doesn't echo the URL.
> 3. At minimum, add `--quiet` to `wget` and clear the output cells before commit.

Thanks, I have added `--quiet` flag, updated the interpolation code and added a changelog line.

**Review by peterdsharpe — APPROVED** [review]
Approving after offline discussion.

One remaining item that we agreed to do before merge - re-running the notebooks to do a) end-to-end integration testing, and b) cleanup outdated output cells that could mislead end-users using this as primary documentation.
